### PR TITLE
gini_impurity function property and test cases for this

### PR DESCRIPTION
**Pull request by az85252**
I have added gini impurity into categorical column, basically likelihood of an incorrect classification of a new instance of a random variable. Please review.

### CODE REVIEW

File: dataprofiler/profilers/categorical_column_profile.py [hunk 1]
```diff
@@ -173,4 +177,15 @@ def update(self, df_series):
 
         self._update_helper(df_series, profile)
 
-        return self
\ No newline at end of file
+        return self
+
+    @property
+    def gini_impurity(self):
```
Comment (AnhTruong) [inline dataprofiler/profilers/categorical_column_profile.py:183]: can you add description and the link to calculate gini impurity?

File: dataprofiler/profilers/categorical_column_profile.py [hunk 2]
```diff
@@ -35,6 +35,7 @@ def __init__(self, name, options=None):
         self._categories = defaultdict(int)
         self.__calculations = {}
         self._filter_properties_w_options(self.__calculations, options)
+        self._gini_impurity = 0
```
Comment (AnhTruong) [inline dataprofiler/profilers/categorical_column_profile.py:38]: should this be None initially, like other stats?

**Review by AnhTruong — COMMENTED** [review]
some comments

**Review by AnhTruong — APPROVED** [review]

File: dataprofiler/profilers/categorical_column_profile.py [hunk 3]
```diff
@@ -81,6 +81,9 @@ def profile(self):
             profile["statistics"].update(
                 dict(categories=self.categories)
             )
+            profile["statistics"].update(
+                dict(gini_impurity=self.gini_impurity)
+            )
```
Comment (ChrisWallace2020) [inline dataprofiler/profilers/categorical_column_profile.py:86]: I believe you could just add the gini_impurity key to the call to update above, with the categories.

File: dataprofiler/tests/profilers/test_categorical_column_profile.py [hunk 4]
```diff
@@ -280,6 +284,28 @@ def test_categorical_merge(self):
         self.assertEqual(report_count['abcd'], expected_dict['abcd'])
         self.assertEqual(report_count['b'], expected_dict['b'])
 
+    def test_gini_impurity(self):
+        # Normal test
+        df_categorical = pd.Series(["y", "y", "y", "y", "n", "n", "n"])
+        profile = CategoricalColumn(df_categorical.name)
+        profile.update(df_categorical)
+        expected_val = ((4 / 7) * (3/7)) + ((4 / 7) * (3/7))
```
Comment (ChrisWallace2020) [inline dataprofiler/tests/profilers/test_categorical_column_profile.py:292]: Make the spacing spacing on (4 / 7) consistent with (3/7)

File: dataprofiler/profilers/categorical_column_profile.py [hunk 5]
```diff
@@ -81,6 +81,7 @@ def profile(self):
             profile["statistics"].update(
                 dict(categories=self.categories)
             )
```
Comment (JGSweets) [inline dataprofiler/profilers/categorical_column_profile.py:83]: not your issue, but if you have to make another change in this pr, could you change this to match the format below:
`profile["statistics"]["categories"] =. self.categories`

File: dataprofiler/profilers/categorical_column_profile.py [hunk 6]
```diff
@@ -173,4 +174,19 @@ def update(self, df_series):
...
         self._update_helper(df_series, profile)
 
-        return self
\ No newline at end of file
+        return self
+
+    @property
+    def gini_impurity(self):
+        """
+        Property for gini impurity
```
Comment (JGSweets) [inline dataprofiler/profilers/categorical_column_profile.py:182]: lets put a reference to the formula or show the formula used here. Also need a space between the description and the return.

File: dataprofiler/profilers/categorical_column_profile.py [hunk 7]
```diff
@@ -173,4 +174,19 @@ def update(self, df_series):
...
+        :return: None or gini impurity probability
+        """
+        if not self._categories:
+            return None
+        summation = 0
+        total = sum(self._categories.values())
```
Comment (JGSweets) [inline dataprofiler/profilers/categorical_column_profile.py:188]: this value may already exist as `self.sample_size`

File: dataprofiler/tests/profilers/test_categorical_column_profile.py [hunk 8]
```diff
@@ -261,13 +262,16 @@ def test_categorical_merge(self):
...
+        report_gini = report['statistics'].pop('gini_impurity')
         expected_profile = dict(
             categorical=True,
             statistics=dict([
                 ('unique_count', 16),
                 ('unique_ratio', 16 / 1000)
             ]),
         )
+        expected_gini = .914600550
+        self.assertAlmostEqual(report_gini, expected_gini)
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_categorical_column_profile.py:274]: is it possible to get the exact math like above or do we have to do almost equal? if we can get it then we won't need to pop it either.

File: dataprofiler/profilers/categorical_column_profile.py (hunk 2 above)
Comment (JGSweets) [reply dataprofiler/profilers/categorical_column_profile.py]: Yes, good catch.

File: dataprofiler/profilers/categorical_column_profile.py [hunk 9]
```diff
@@ -173,4 +172,26 @@ def update(self, df_series):
...
+        G = Σ(i=1; J): P(i) * (1 - P(i)), where i is the category classes.
+        We are traversing through categories and calculating with the column
+
+        :return: None or Gini Impurity probability
+        """
+        if self.sample_size == 0:
+            return None
+        summation = 0
+        #total = sum(self._categories.values())
+        print(self.sample_size)
```
Comment (JGSweets) [inline dataprofiler/profilers/categorical_column_profile.py:193]: remove the print

File: dataprofiler/profilers/categorical_column_profile.py [hunk 10]
```diff
@@ -173,4 +172,26 @@ def update(self, df_series):
...
+        a random variable.
+
+        G = Σ(i=1; J): P(i) * (1 - P(i)), where i is the category classes.
+        We are traversing through categories and calculating with the column
+
+        :return: None or Gini Impurity probability
+        """
+        if self.sample_size == 0:
+            return None
+        summation = 0
```
Comment (JGSweets) [inline dataprofiler/profilers/categorical_column_profile.py:191]: Maybe just call this `gini_impurity`? thoughts?

**Review by JGSweets — APPROVED** [review]

**Review by AnhTruong — APPROVED** [review]

**Pull request merged**